tikv/raftstore/util.go

```diff
@@ -0,0 +1,229 @@
+			l.boundSuspect = nil
+			l.boundValid = &bound
+		}
+		if l.boundValid != nil && l.boundValid.Before(bound) {
+			l.boundValid = &bound
+		}
+	} else {
+		// Or an empty lease
+		l.boundValid = &sendTs
+	}
+
+	// Renew remote if it's valid.
+	if l.boundValid != nil {
+		if l.boundValid.Sub(l.lastUpdate) > l.maxDrift {
+			l.lastUpdate = *l.boundValid
+			if l.remote != nil {
+				l.remote.Renew(*l.boundValid)
+			}
+		}
+	}
+}
+
+/// Suspect the lease to the bound.
+func (l *Lease) Suspect(sendTs time.Time) {
+	l.ExpireRemoteLease()
+	bound := l.nextExpiredTime(sendTs)
+	l.boundValid = nil
+	l.boundSuspect = &bound
+}
+
+/// Inspect the lease state for the ts or now.
+func (l *Lease) Inspect(ts *time.Time) LeaseState {
+	if l.boundSuspect != nil {
+		return LeaseState_Suspect
+	}
+	if l.boundValid != nil {
+		if ts == nil {
+			t := time.Now()
+			ts = &t
+		}
+		if ts.Before(*l.boundValid) {
+			return LeaseState_Valid
+		} else {
+			return LeaseState_Expired
+		}
+	}
+	return LeaseState_Expired
+}
+
+func (l *Lease) Expire() {
+	l.ExpireRemoteLease()
+	l.boundValid = nil
+	l.boundSuspect = nil
+}
+
+func (l *Lease) ExpireRemoteLease() {
+	// Expire remote lease if there is any.
+	if l.remote != nil {
+		l.remote.Expire()
+		l.remote = nil
+	}
+}
+
+/// Return a new `RemoteLease` if there is none.
+func (l *Lease) MaybeNewRemoteLease(term uint64) *RemoteLease {
+	if l.remote != nil {
+		if l.remote.term == term {
+			// At most one connected RemoteLease in the same term.
+			return nil
+		} else {
+			// Term has changed. It is unreachable in the current implementation,
+			// because we expire remote lease when leaders step down.
+			panic("Must expire the old remote lease first!")
+		}
+	}
+	expiredTime := uint64(0)
+	if l.boundValid != nil {
+		expiredTime = TimeToU64(*l.boundValid)
+	}
+	remote := &RemoteLease {
+		expiredTime: &expiredTime,
+		term: term,
+	}
+	// Clone the remote.
+	remoteClone := &RemoteLease {
+		expiredTime: &expiredTime,
+		term: term,
+	}
+	l.remote = remote
+	return remoteClone
+}
+
+/// A remote lease, it can only be derived by `Lease`. It will be sent
+/// to the local read thread, so name it remote. If Lease expires, the remote must
+/// expire too.
+type RemoteLease struct {
+	expiredTime *uint64
+	term uint64
+}
+
+func (r *RemoteLease) Inspect(ts *time.Time) LeaseState {
+	expiredTime := atomic.LoadUint64(r.expiredTime)
+	if ts == nil {
+		t := time.Now()
+		ts = &t
+	}
+	if ts.Before(U64ToTime(expiredTime)) {
+		return LeaseState_Valid
+	} else {
+		return LeaseState_Expired
+	}
+}
+
+func (r *RemoteLease) Renew(bound time.Time) {
+	atomic.StoreUint64(r.expiredTime, TimeToU64(bound))
+}
+
+func (r *RemoteLease) Expire() {
+	atomic.StoreUint64(r.expiredTime, 0)
+}
+
+func (r *RemoteLease) Term() uint64 {
+	return r.term
+}
+
+const (
+	NSEC_PER_MSEC uint64 = 1000000
+	SEC_SHIFT uint64 = 10
+	MSEC_MASK uint64 = (1 << SEC_SHIFT) - 1
+)
+
+func TimeToU64(t time.Time) uint64 {
+	sec := uint64(t.Unix())
```

Comment:
We can use `t.UnixNano()`.

Reply:
I think it should be the nanoseconds in current seconds.

Reply:
In Go source
```
func (t Time) UnixNano() int64 {
	return (t.unixSec())*1e9 + int64(t.nsec())
}
```

Reply:
What i meant is i need the nanosecond offset within current second.